From ead108712dd560fb46d10535a28a0cfca496b32a Mon Sep 17 00:00:00 2001
From: TSKangetsu <x403070567@live.com>
Date: Thu, 19 Mar 2026 14:36:34 +0800
Subject: [PATCH] arm64: dts: rockchip: fix performance scaling and power
 models for acropi\n\n- Increase PMIC voltage limits to support higher OPP
 points.\n- Fix CPU clocks for all cores.\n- Link NVMEM leakage cells to
 enable hardware telemetry.\n- Add missing power-model properties for Mali
 GPU.\n\nUpstream-Status: Inappropriate [configuration]\nSigned-off-by:
 TSKangetsu <x403070567@live.com>

---
 .../boot/dts/rockchip/rk3566-acropi-lp4x.dts  | 108 +++++++++++++++---
 1 file changed, 94 insertions(+), 14 deletions(-)

diff --git a/arch/arm64/boot/dts/rockchip/rk3566-acropi-lp4x.dts b/arch/arm64/boot/dts/rockchip/rk3566-acropi-lp4x.dts
index 8b74f31c1a7e..7d139e413458 100644
--- a/arch/arm64/boot/dts/rockchip/rk3566-acropi-lp4x.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3566-acropi-lp4x.dts
@@ -51,8 +51,29 @@ vcc3v3_sys: vcc3v3-sys {
                 vin-supply = <&vcc5v0_sys>;
         };
 
+        cpu_opp_table_fixed: opp-table-cpu {
+                compatible = "operating-points-v2";
+                nvmem-cells = <&cpu_leakage>;
+                nvmem-cell-names = "leakage";
+                rockchip,grf = <&grf>;
+                rockchip,pvtpll = <&grf>;
+                rockchip,pvtm-voltage-sel = <0>;
+
+                opp-408000000 { opp-hz = /bits/ 64 <408000000>; opp-microvolt = <900000>; };
+                opp-600000000 { opp-hz = /bits/ 64 <600000000>; opp-microvolt = <900000>; };
+                opp-816000000 { opp-hz = /bits/ 64 <816000000>; opp-microvolt = <900000>; };
+                opp-1008000000 { opp-hz = /bits/ 64 <1008000000>; opp-microvolt = <1000000>; };
+                opp-1200000000 { opp-hz = /bits/ 64 <1200000000>; opp-microvolt = <1100000>; };
+        };
+
         gpu_opp_table_fixed: opp-table-gpu {
                 compatible = "operating-points-v2";
+                nvmem-cells = <&gpu_leakage>;
+                nvmem-cell-names = "leakage";
+                rockchip,grf = <&grf>;
+                rockchip,pvtpll = <&grf>;
+                rockchip,pvtm-voltage-sel = <0>;
+
                 opp-200000000 { opp-hz = /bits/ 64 <200000000>; opp-microvolt = <900000>; };
                 opp-400000000 { opp-hz = /bits/ 64 <400000000>; opp-microvolt = <900000>; };
                 opp-600000000 { opp-hz = /bits/ 64 <600000000>; opp-microvolt = <1000000>; };
@@ -61,6 +82,12 @@ gpu_opp_table_fixed: opp-table-gpu {
 
         npu_opp_table_fixed: opp-table-npu {
                 compatible = "operating-points-v2";
+                nvmem-cells = <&npu_leakage>;
+                nvmem-cell-names = "leakage";
+                rockchip,grf = <&grf>;
+                rockchip,pvtpll = <&grf>;
+                rockchip,pvtm-voltage-sel = <0>;
+
                 opp-200000000 { opp-hz = /bits/ 64 <200000000>; opp-microvolt = <900000>; };
                 opp-400000000 { opp-hz = /bits/ 64 <400000000>; opp-microvolt = <900000>; };
                 opp-600000000 { opp-hz = /bits/ 64 <600000000>; opp-microvolt = <1000000>; };
@@ -68,21 +95,36 @@ npu_opp_table_fixed: opp-table-npu {
 };
 
 &cpu0 {
-       clocks = <&cru ARMCLK>;
-       cpu-supply = <&vdd_logic>;
+        clocks = <&cru ARMCLK>;
+        operating-points-v2 = <&cpu_opp_table_fixed>;
+        cpu-supply = <&vdd_logic>;
 };
 
-&cpu1 { cpu-supply = <&vdd_logic>; };
-&cpu2 { cpu-supply = <&vdd_logic>; };
-&cpu3 { cpu-supply = <&vdd_logic>; };
+&cpu1 {
+        clocks = <&cru ARMCLK>;
+        operating-points-v2 = <&cpu_opp_table_fixed>;
+        cpu-supply = <&vdd_logic>;
+};
+
+&cpu2 {
+        clocks = <&cru ARMCLK>;
+        operating-points-v2 = <&cpu_opp_table_fixed>;
+        cpu-supply = <&vdd_logic>;
+};
+
+&cpu3 {
+        clocks = <&cru ARMCLK>;
+        operating-points-v2 = <&cpu_opp_table_fixed>;
+        cpu-supply = <&vdd_logic>;
+};
 
 &usb2phy0 {
-       status = "okay";
-       phy-supply = <&vcc5v0_sys>;
+        status = "okay";
+        phy-supply = <&vcc5v0_sys>;
 };
 
 &u2phy0_otg {
-       status = "okay";
+        status = "okay";
 };
 
 &usbdrd_dwc3 {
@@ -95,7 +137,7 @@ &usbdrd_dwc3 {
 };
 
 &usbdrd30 {
-       status = "okay";
+        status = "okay";
 };
 
 &scmi {
@@ -115,6 +157,7 @@ pmic: pmic@20 {
                 pinctrl-names = "default";
                 pinctrl-0 = <&pmic_int_l>;
                 rockchip,system-power-controller;
+                rockchip,vcc13-feedback-soc;
                 wakeup-source;
 
                 vcc1-supply = <&vcc3v3_sys>;
@@ -128,16 +171,45 @@ pmic: pmic@20 {
                 vcc9-supply = <&vcc3v3_sys>;
 
                 regulators {
-                        vdd_logic: DCDC_REG1 { regulator-name = "vdd_logic"; regulator-always-on; regulator-boot-on; };
-                        vdd_gpu: DCDC_REG2 { regulator-name = "vdd_gpu"; regulator-always-on; regulator-boot-on; };
-                        vcc_ddr: DCDC_REG3 { regulator-name = "vcc_ddr"; regulator-always-on; regulator-boot-on; };
-                        vdd_npu: DCDC_REG4 { regulator-name = "vdd_npu"; regulator-always-on; regulator-boot-on; };
+                        vdd_logic: DCDC_REG1 {
+                                regulator-name = "vdd_logic";
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <500000>;
+                                regulator-max-microvolt = <1350000>;
+                        };
+
+                        vdd_gpu: DCDC_REG2 {
+                                regulator-name = "vdd_gpu";
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <500000>;
+                                regulator-max-microvolt = <1350000>;
+                        };
+
+                        vcc_ddr: DCDC_REG3 {
+                                regulator-name = "vcc_ddr";
+                                regulator-always-on;
+                                regulator-boot-on;
+                        };
+
+                        vdd_npu: DCDC_REG4 {
+                                regulator-name = "vdd_npu";
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <500000>;
+                                regulator-max-microvolt = <1350000>;
+                        };
                 };
         };
 };
 
 &pinctrl {
-        pmic { pmic_int_l: pmic-int-l { rockchip,pins = <0 RK_PA3 RK_FUNC_GPIO &pcfg_pull_up>; }; };
+        pmic {
+                pmic_int_l: pmic-int-l {
+                        rockchip,pins = <0 RK_PA3 RK_FUNC_GPIO &pcfg_pull_up>;
+                };
+        };
 };
 
 &rknpu {
@@ -155,6 +227,14 @@ &gpu {
         operating-points-v2 = <&gpu_opp_table_fixed>;
         mali-supply = <&vdd_gpu>;
         status = "okay";
+
+        gpu_power_model: power-model {
+                compatible = "arm,mali-simple-power-model";
+                static-coefficient = <411>;
+                dynamic-coefficient = <733>;
+                ts = <32000 4700 (-80) 2>;
+                thermal-zone = "gpu-thermal";
+        };
 };
 
 &uart2 {
